File: src/main/java/com/nerdearla/workshop/service/PaymentsService.java
package com.nerdearla.workshop.service;

import com.nerdearla.workshop.dto.payment.PaymentResponse;
import com.nerdearla.workshop.model.PaymentOperation;
import com.nerdearla.workshop.model.User;
import org.springframework.stereotype.Service;

@Service
public class PaymentsService {

    private final PaymentIdProvider paymentIdProvider;
    private final UserService userService;
    private final PaymentMethodService paymentMethodService;
    private final FraudService fraudService;
    private final GatewayService gatewayService;
    private final BuyerService buyerService;
    private final SellerService sellerService;

    public PaymentsService(
            PaymentIdProvider paymentIdProvider,
            UserService userService,
            PaymentMethodService paymentMethodService,
            FraudService fraudService,
            GatewayService gatewayService,
            BuyerService buyerService,
            SellerService sellerService
    ) {
        this.paymentIdProvider = paymentIdProvider;
        this.userService = userService;
        this.paymentMethodService = paymentMethodService;
        this.fraudService = fraudService;
        this.gatewayService = gatewayService;
        this.buyerService = buyerService;
        this.sellerService = sellerService;
    }

    public PaymentResponse processPayment(PaymentOperation operation) {
        String id = paymentIdProvider.getNext();
        operation.setPaymentId(id);

        // validar qr_id?

        // validar amount e installments?

        User buyer = userService.findValidUser(operation.getPaymentRequest().getBuyerId());
        operation.setBuyer(buyerService.findBuyer(buyer.getId()));

        // Obtener paymentMethod y agregarlo a operation?
        paymentMethodService.authorize(operation.getPaymentRequest().getPaymentMethod());

        User seller = userService.findValidUser(operation.getPaymentRequest().getSellerId());
        operation.setSeller(sellerService.findSeller(seller.getId()));

        fraudService.authorize(operation);

        gatewayService.authorize(operation);

        return new PaymentResponse();
    }
}

File: src/main/java/com/nerdearla/workshop/service/PaymentsService.kt
package com.nerdearla.workshop.service

import com.nerdearla.workshop.dto.payment.PaymentRequest.buyerId
import com.nerdearla.workshop.dto.payment.PaymentRequest.paymentMethod
import com.nerdearla.workshop.dto.payment.PaymentRequest.sellerId
import com.nerdearla.workshop.model.Buyer
import com.nerdearla.workshop.model.Seller
import com.nerdearla.workshop.dto.payment.PaymentRequest
import com.nerdearla.workshop.model.PaymentOperation
import com.nerdearla.workshop.service.PaymentIdProvider
import com.nerdearla.workshop.service.UserService
import com.nerdearla.workshop.service.PaymentMethodService
import com.nerdearla.workshop.service.FraudService
import com.nerdearla.workshop.service.GatewayService
import com.nerdearla.workshop.dto.payment.PaymentResponse
import com.nerdearla.workshop.dto.payment.PaymentMethod
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RequestMapping
import com.nerdearla.workshop.service.PaymentsService
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import javax.validation.Valid
import org.springframework.boot.autoconfigure.SpringBootApplication
import kotlin.jvm.JvmStatic
import org.springframework.boot.SpringApplication
import org.springframework.stereotype.Service

@Service
class PaymentsService(
    private val paymentIdProvider: PaymentIdProvider,
    private val userService: UserService,
    private val paymentMethodService: PaymentMethodService,
    private val fraudService: FraudService,
    private val gatewayService: GatewayService
) {
    fun processPayment(paymentRequest: PaymentRequest?): PaymentResponse {
        val id = paymentIdProvider.next
        userService.findValidUser(paymentRequest!!.buyerId)
        paymentMethodService.authorize(paymentRequest.paymentMethod)
        userService.findValidUser(paymentRequest.sellerId)
        val operation = PaymentOperation()
        fraudService.authorize(operation)
        gatewayService.authorize(operation)
        return PaymentResponse()
    }
}